src/test/scala/chiselTests/BoringUtilsSpec.scala

```diff
@@ -302,4 +303,170 @@ class BoringUtilsSpec extends ChiselFlatSpec with ChiselRunners with Utils with
   }
+
+  "Downwards tap from parent to child" should "work" in {
+    val chirrtl = circt.stage.ChiselStage.emitCHIRRTL(
+      new RawModule {
+        class Foo() extends RawModule {
+          val internalWire = Wire(Bool())
+        }
+
+        val foo = Module(new Foo())
+        val outProbe = IO(probe.Probe(Bool()))
+        val out = IO(Bool())
+
+        probe.define(outProbe, BoringUtils.tap(foo.internalWire))
+
+        out := BoringUtils.tapAndRead(foo.internalWire)
+      },
+      Array("--full-stacktrace", "--throw-on-first-error")
+    )
+    matchesAndOmits(chirrtl)(
+      "output bore : Probe<UInt<1>>",
+      "output out_bore : Probe<UInt<1>>",
+      "define bore = probe(internalWire)",
+      "define out_bore = probe(internalWire)",
+      "define outProbe = foo.bore",
+      "out <= read(foo.out_bore)"
+    )()
+  }
+
+  "Downwards tap from grandparent to grandchild" should "work" in {
+    val chirrtl = circt.stage.ChiselStage.emitCHIRRTL(
+      new RawModule {
+        class Bar() extends RawModule {
+          val internalWire = Wire(Bool())
+        }
+
+        class Foo() extends RawModule {
+          val bar = Module(new Bar())
+        }
+
+        val foo = Module(new Foo())
+        val out = IO(Bool())
+
+        out := BoringUtils.tapAndRead(foo.bar.internalWire)
+      },
+      Array("--full-stacktrace", "--throw-on-first-error")
+    )
+    matchesAndOmits(chirrtl)(
+      "output out_bore : Probe<UInt<1>>",
+      "define out_bore = probe(internalWire)",
+      "output out_bore : Probe<UInt<1>>",
+      "define out_bore = bar.out_bore",
+      "out <= read(foo.out_bore)"
+    )()
+  }
+
+  "Upwards tap from child to parent" should "work" in {
+    val chirrtl = circt.stage.ChiselStage.emitCHIRRTL(
+      new RawModule {
+        class Foo(parentData: Data) extends RawModule {
+          val outProbe = IO(probe.Probe(Bool()))
+          val out = IO(Bool())
+          probe.define(outProbe, BoringUtils.tap(parentData))
+          out := BoringUtils.tapAndRead(parentData)
+        }
+
+        val parentWire = Wire(Bool())
+
+        val foo = Module(new Foo(parentWire))
+      },
+      Array("--full-stacktrace", "--throw-on-first-error")
+    )
+    matchesAndOmits(chirrtl)(
+      "input bore : UInt<1>",
+      "define outProbe = probe(bore)",
+      "out <= read(probe(out_bore))",
+      "foo.bore <= parentWire",
+      "foo.out_bore <= parentWire"
+    )()
+  }
+
+  "Upwards tap from grandchild to grandparent" should "work" in {
+    val chirrtl = circt.stage.ChiselStage.emitCHIRRTL(
+      new RawModule {
+        class Bar(grandParentData: Data) extends RawModule {
+          val out = IO(Bool())
+          out := BoringUtils.tapAndRead(grandParentData)
+        }
+
+        class Foo(parentData: Data) extends RawModule {
+          val out = IO(Bool())
+          val bar = Module(new Bar(parentData))
+          out := bar.out
```

Comment:
Nit: the IO on the parent are unneeded:
```suggestion
          val bar = Module(new Bar(parentData))
```